### PR TITLE
TDL-16246: Make MAX_API_RESPONSE_SIZE configurable and add pagination for card stream.

### PR DESCRIPTION
**Pull request by savan-chovatiya**
# Description of change
[TDL-16246](https://jira.talendforge.org/browse/TDL-16246): Make MAX_API_RESPONSE_SIZE configurable
- Added support of configurable  parameter `cards_response_size`  in config file for card stream.(Default : 1000)
- Added pagination for the card stream.

# Manual QA steps
 - The verified output of the old tap with data emitted with a new approach for card stream.
- Provided limit in integer, and string format and verified that it's used in the request.

 
# Risks
 - Minimal. Test coverage suggests no issues with pagination.
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_trello/streams.py [hunk 1]
```diff
@@ -373,8 +373,52 @@ class Cards(AddCustomFields, ChildStream):
     key_properties = ["id"]
     replication_method = "FULL_TABLE"
     parent_class = Boards
-    MAX_API_RESPONSE_SIZE = 20000
-    params = {'limit': 20000, 'customFieldItems': 'true'}
+    MAX_API_RESPONSE_SIZE = 1000
+
+    def get_records(self, format_values, additional_params=None):
+        # Get max_api_response_size from config and set to paramater
```
Comment (dbshah1212) [inline tap_trello/streams.py:379]:
```suggestion
        # Get max_api_response_size from config and set to parameter
```

File: tap_trello/streams.py [hunk 2]
```diff
@@ -373,8 +373,52 @@ class Cards(AddCustomFields, ChildStream):
...
+            self.MAX_API_RESPONSE_SIZE = int(response_size)
+        self.params = {'limit': self.MAX_API_RESPONSE_SIZE, 'customFieldItems': 'true'}
+
+        # Set window_end with current time
+        window_end = utils.strftime(utils.now())
+
+        custom_fields_map, dropdown_options_map = self.build_custom_fields_maps(parent_id_list=format_values)
+        while True:
+
+            # Get records for cards before specified time or card ID
```
Comment (dbshah1212) [inline tap_trello/streams.py:391]: Can explain in a comment, why we can use card ID instead of time?

File: tap_trello/streams.py [hunk 3]
```diff
@@ -373,8 +373,52 @@ class Cards(AddCustomFields, ChildStream):
...
+            # If records are same as limit then shift window to get older data
+            if len(records) == self.MAX_API_RESPONSE_SIZE:
+                LOGGER.info("%s - Collected  %s records for board %s.",
+                            self.stream_id,
+                            len(records),
+                            format_values[0])
+
+                # Sort cards based on card ID as API returns latest records but in unorder manner
+                records = sorted(records, key=lambda x: x['id'])
+                # API returns latest records so set sub_window_end to smallest card id to get older data
```
Comment (dbshah1212) [inline tap_trello/streams.py:406]: window_end instead of sub_window_end, right?

**Review by dbshah1212 — COMMENTED** [review]
Looks good, provided some comment changes.

File: tap_trello/__init__.py [hunk 4]
```diff
@@ -3,7 +3,7 @@
 from singer.catalog import Catalog, write_catalog
 from tap_trello.discover import do_discover
 from tap_trello.sync import do_sync
-import tap_trello.streams as streams
+from tap_trello import streams
```
Comment (savan-chovatiya) [inline tap_trello/__init__.py:6]: The updated above thing to fix pylint error.

File: tap_trello/streams.py (hunk 2 above)
Comment (savan-chovatiya) [reply tap_trello/streams.py:391]: Yes, I added details and a reference article link for the same.

File: tap_trello/streams.py (hunk 1 above)
Comment (savan-chovatiya) [reply tap_trello/streams.py]: Done changes.

File: tap_trello/streams.py (hunk 3 above)
Comment (savan-chovatiya) [reply tap_trello/streams.py]: Yes, updated the sub_window_end to window_end.

File: tests/test_trello_pagination.py [hunk 5]
```diff
@@ -206,6 +212,12 @@ def test_run(self):
 
         for stream in self.testable_streams():
             with self.subTest(stream=stream):
+        
+                # Page limit for card stream(50) is configured in config above
+                self.API_LIMIT = 50 if stream == "cards" else 1000
```
Comment (kspeer825) [inline tests/test_trello_pagination.py:217]: This is set on line 132 already.

**Review by kspeer825 — APPROVED** [review]
Approved with 1 comment.

File: tests/test_trello_pagination.py (hunk 5 above)
Comment (savan-chovatiya) [reply tests/test_trello_pagination.py:217]: Yes, but both are under different for loop and we are updating API_LIMIT on for loop iteration so need to set it again.

Comment (kspeer825) [reply tests/test_trello_pagination.py:217]: A more consistent implementation would be adding API LIMIT as a piece of expected metadata in base.

Comment (savan-chovatiya) [reply tests/test_trello_pagination.py:217]: Yes, that's true but tests written in this tap are not in that format. Every test has its own methods for retrieving expected things and no expected metadata functions and also the pagination test is only for actions and card streams.

**Review by KrisPersonal — APPROVED** [review]

**Review by luandy64 — APPROVED** [review]

**Pull request merged**